base/event.jl

```diff
@@ -343,3 +343,50 @@ function Timer(cb::Function, timeout::Real, repeat::Real=0.0)
     yieldto(waiter)
     return t
 end
+
+type TimeoutException <: Exception
+    msg::AbstractString
+
+    TimeoutException() = new("")
+    TimeoutException(msg) = new(msg)
+end
+
+"""
+    @timeout(t, ex, [msg])
+
+Set a timeout of `t` seconds on the execution time of expression `ex`.
+Throws `TimeoutException` with optional error message `msg`
+if the timeout is triggered.
+
+Be aware that the exception raised by the timeout will not abort
+the evaluation of `ex`.
```

Comment:
No, It can simply be caused by the user function not hitting a `yield` at all.
This is also why I'm wondering how useful this function is.
I believe a timeout is usually used to make blocking code relatively non-blocking and to abort operation that is taking too long. This macro doesn't seems to provide any of these.

1. It can only provide feedback (of whether there's a timeout) for operations that are already async (i.e. it has to yield to other tasks)
2. It does not abort any operations.

(1) basically can't be solved with tasks.
For (2), it would be more useful if `sleep(timeout)` task runs a user defined callback instead, which at least has some chance to be able to abort the operation. However, this would remove the majority of the code in this PR and I feel like a `@schedule` with an optional delay might be more useful....